piker/brokers/kucoin.py

```diff
@@ -0,0 +1,668 @@
+# Copyright (C) Jared Goldman (in stewardship for pikers)
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+'''
+Kucoin broker backend
+
+'''
+
+from random import randint
+from typing import Any, Callable, Optional, Literal, AsyncGenerator
+from contextlib import asynccontextmanager as acm
+from datetime import datetime
+import time
+import math
+import base64
+import hmac
+import hashlib
+import wsproto
+from uuid import uuid4
+
+import asks
+import tractor
+import trio
+from trio_typing import TaskStatus
+from fuzzywuzzy import process as fuzzy
+import pendulum
+import numpy as np
+
+from piker._cacheables import open_cached_client
+from piker.log import get_logger
+from ._util import DataUnavailable
+from piker.pp import config
+from ..data.types import Struct
+from ..data._web_bs import (
+    open_autorecon_ws,
+    NoBsWs,
+)
+
+log = get_logger(__name__)
+
+_ohlc_dtype = [
+    ('index', int),
+    ('time', int),
+    ('open', float),
+    ('high', float),
+    ('low', float),
+    ('close', float),
+    ('volume', float),
+    ('bar_wap', float),  # will be zeroed by sampler if not filled
+]
+
+
+class KucoinMktPair(Struct, frozen=True):
+    '''
+    Kucoin's pair format
+
+    '''
+
```

Comment:
i tend to *not* prefer having an extra line between `'''` and docstring since we already put one as last content line in doc string.

Reply:
Done